### ISSUE TITLE
Results were inconsistent for each run

### ISSUE THREAD

**Issue by Dengshunge** [huggingface/diffusers#5523]
### Describe the bug

When i run the same code, the sd result is slight different each time, even if set the random seed.  
```
import os
import torch
import numpy as np
import cv2
import random

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

from diffusers import StableDiffusionPipeline
from diffusers import DDIMScheduler
from diffusers import DDPMPipeline

model_id = "./stable-diffusion-v1-5"
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float32, revision="fp32")
# pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
pipe = pipe.to("cuda")
generator = torch.Generator(device="cuda").manual_seed(seed)

prompt = "a photo of an astronaut riding a horse on mars"
image = pipe(prompt,num_inference_steps=10,generator=generator).images[0]  

image = np.array(image)
print(image.sum())
```

I found that `https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention_processor.py#L1044` this code case different results. When i change `dropout_p=-1`, the result is same for each run. Maybe the op `scaled_dot_product_attention` have some problem.

```
import os
import torch
import numpy as np
import cv2
import random
import torch.nn.functional as F

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

query = torch.rand((2, 8, 4096, 40)).float().cuda()
key = torch.rand((2, 8, 4096, 40)).float().cuda()
value = torch.rand((2, 8, 4096, 40)).float().cuda()

hidden_states1 = F.scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=-1, is_causal=False) 
hidden_states2 = F.scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=-1, is_causal=False)

print((hidden_states1-hidden_states2).sum()) # when set dropout_p=-1, the diff = 0
```

### Reproduction

```
import os
import torch
import numpy as np
import cv2
import random

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

from diffusers import StableDiffusionPipeline
from diffusers import DDIMScheduler
from diffusers import DDPMPipeline

model_id = "./stable-diffusion-v1-5"
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float32, revision="fp32")
# pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
pipe = pipe.to("cuda")
generator = torch.Generator(device="cuda").manual_seed(seed)

prompt = "a photo of an astronaut riding a horse on mars"
image = pipe(prompt,num_inference_steps=10,generator=generator).images[0]  

image = np.array(image)
print(image.sum())
```

### System Info

- `diffusers` version: 0.22.0.dev0
- Platform: Linux-5.4.0-144-generic-x86_64-with-glibc2.27
- Python version: 3.10.13
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Huggingface_hub version: 0.17.3
- Transformers version: 4.34.0
- Accelerate version: 0.23.0
- xFormers version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

**Comment by patrickvonplaten** [comment]
cc @DN6

**Comment by DN6** [comment]
@Dengshunge I'm unable to reproduce this with the code example provided. 

In general, bitwise-identical results aren't expected when running on GPU even when setting `torch.backends.cudnn.deterministic = True`(see: https://github.com/openai/baselines/issues/805)

A couple of ways to mitigate this in diffusers (although it might not fully solve the issue) 

Use cpu with the generator `torch.Generator("cpu").manual_seed(seed)` (under the hood, diffusers moves the tensor to GPU after creation)

Scaled Dot Product Attention uses Flash Atteention which is also slightly more non-deterministic than regular attention.  
https://discuss.pytorch.org/t/scaled-dot-product-attn-not-giving-exact-same-results-as-normal-attention/176040

You can use the default UNet attention processor to `pipe.unet.set_default_attn_processor` to avoid using Flash Attention.
